tests/unit/via/services/checkmate_test.py

```diff
@@ -0,0 +1,133 @@
+        )
+
+    def test_raise_if_blocked_raises_if_the_given_url_is_blocked(
+        self, checkmate_client, svc, block_response
+    ):
+        checkmate_client.check_url.return_value = block_response
+
+        with pytest.raises(HTTPTemporaryRedirect) as exc_info:
+            svc.raise_if_blocked(sentinel.url)
+
+        assert exc_info.value.location == block_response.presentation_url
+
+    def test_raise_if_blocked_raises_if_the_given_url_is_bad(
+        self, checkmate_client, svc
+    ):
+        checkmate_client.check_url.side_effect = CheckmateBadURL()
+
+        with pytest.raises(BadURL) as exc_info:
+            svc.raise_if_blocked(sentinel.url)
+
+        assert exc_info.value.args[0] == checkmate_client.check_url.side_effect
+        assert exc_info.value.url == sentinel.url
+
+    def test_raise_if_blocked_doesnt_raise_if_checkmate_crashes(
+        self, checkmate_client, svc
+    ):
+        checkmate_client.check_url.side_effect = CheckmateException
+
+        svc.raise_if_blocked(sentinel.url)
+
+    @pytest.fixture
+    def block_response(self):
+        return create_autospec(BlockResponse, instance=True, spec_set=True)
+
+    @pytest.fixture
+    def checkmate_client(self, CheckmateClient):
+        return CheckmateClient.return_value
+
+    @pytest.fixture
+    def svc(self, checkmate_client):
+        return CheckmateService(
+            checkmate_client=checkmate_client,
+            allow_all=sentinel.allow_all,
+            blocked_for=sentinel.blocked_for,
+            ignore_reasons=sentinel.ignore_reasons,
+        )
+
+
+class TestFactory:
+    def test_it(self, pyramid_request, CheckmateClient, CheckmateService):
+        svc = factory(sentinel.context, pyramid_request)
+
+        CheckmateClient.assert_called_once_with(
+            host=sentinel.url, api_key=sentinel.api_key
+        )
+        CheckmateService.assert_called_once_with(
+            checkmate_client=CheckmateClient.return_value,
+            allow_all=sentinel.allow_all,
+            blocked_for=None,
+            ignore_reasons=sentinel.ignore_reasons,
+        )
+        assert svc == CheckmateService.return_value
+
+    def test_it_passes_the_via_blocked_for_query_param_to_CheckmateService(
+        self, pyramid_request, CheckmateService
+    ):
+        pyramid_request.params["via.blocked_for"] = sentinel.blocked_for
+
+        factory(sentinel.context, pyramid_request)
+
+        assert CheckmateService.call_args[1]["blocked_for"] == sentinel.blocked_for
+
+    @pytest.fixture
+    def pyramid_request(self, pyramid_request):
+        pyramid_request.registry.settings["checkmate_url"] = sentinel.url
+        pyramid_request.registry.settings["checkmate_api_key"] = sentinel.api_key
+        pyramid_request.registry.settings["checkmate_allow_all"] = sentinel.allow_all
+        pyramid_request.registry.settings[
+            "checkmate_ignore_reasons"
+        ] = sentinel.ignore_reasons
+        return pyramid_request
```

Comment:
```suggestion
    @pytest.mark.parametrize("blocked_for", (None, sentinel.blocked_for))
    def test_it(self, pyramid_request, CheckmateClient, CheckmateService, blocked_for):
        pyramid_request.registry.settings = {
            "checkmate_url": sentinel.url,
            "checkmate_api_key": sentinel.api_key,
            "checkmate_allow_all": sentinel.allow_all,
            "checkmate_ignore_reasons": sentinel.ignore_reasons,
        }
        if blocked_for:
            pyramid_request.params["via.blocked_for"] = sentinel.blocked_for

        svc = factory(sentinel.context, pyramid_request)

        CheckmateClient.assert_called_once_with(
            host=sentinel.url, api_key=sentinel.api_key
        )
        CheckmateService.assert_called_once_with(
            checkmate_client=CheckmateClient.return_value,
            allow_all=sentinel.allow_all,
            blocked_for=blocked_for,
            ignore_reasons=sentinel.ignore_reasons,
        )
        assert svc == CheckmateService.return_value
```

I find the back and forth between the fixture, sentinels and tests a little complicated. I think we can do this as above in a oner.

Another option might be to assert with statements similar to we do with `blah.return_value`:

```python
assert blah == pyramid_request.registry.settings["checkmate_url"]

# vs.

assert blah == sentinel.checkmate_url
```

Reply:
The registry settings etc are in a fixture because there's more than one test for the factory, so if you don't put the set up in a shared fixture you end up having to duplicate it in each test.

I actually don't mind the pattern of putting test _setup_ in fixtures but then putting the test logic itself in the tests: it nicely separates setup and test and can make the test logic a lot clearer because the tests are stuffed full of distracting setup code. As long as it's done clearly with simple fixtures and with the fixtures sticking to setup and not containing test logic.

And I think when using clear sentinel values it's fairly obvious:

```python
CheckmateService.assert_called_once_with(
    ...
    allow_all=sentinel.allow_all_setting,
    ...
)
```

You may not be able to see exactly where it comes from without scanning the rest of the test class for `sentinel.allow_all_setting`, but doesn't the name alone makes it pretty clear that this is a sentinel value standing in for the `allow_all` setting? And it's eminently greppable too.

(This suggests that maybe longer names like `sentinel.allow_all_setting` might be better than just `sentinel.allow_all`, but the less-specific `sentinel.allow_all` name is one that wouldn't have to change if we e.g. moved the value from a setting to a request param or something.)

Reply:
>so if you don't put the set up in a shared fixture you end up having to duplicate it in each test.

The suggestion above replaces the two tests with a single parameterized test